simplemma/language_detector.py

```diff
@@ -87,42 +76,102 @@ def lang_detector(
     greedy: bool = False,
     dictionary_factory: DictionaryFactory = DictionaryFactory(),
     token_sampler: TokenSampler = TokenSampler(),
-    backup_sampler: TokenSampler = RelaxedTokenSampler(),
+    backup_sampler: TokenSampler = TokenSampler(
+        tokenizer=Tokenizer(RELAXED_SPLIT_INPUT),
+        max_tokens=1000,
+        capitalized_threshold=0,
+    ),
 ) -> List[Tuple[str, float]]:
-    """Determine which proportion of the text is in the target language(s).
-    Perform a first run and further discriminate between the results if necessary."""
-    myresults = {}  # Dict[str, float]
-    tokens = token_sampler.sample_tokens(text)
-    total_tokens = len(tokens)
-    if total_tokens == 0:
-        return _return_default()
-    # iterate
-    dictionaries = dictionary_factory.get_dictionaries(lang)
-    for lang_code, lang_dictionary in dictionaries.items():
-        in_target = 0
-        for token in tokens:
-            candidate = _return_lemma(
-                token, lang_dictionary, greedy=greedy, lang=lang_code
-            )
-            if candidate is not None:
-                in_target += 1
-        # compute results
-        found_ratio = in_target / total_tokens
-        myresults[lang_code] = found_ratio
-        unknown = 1 - found_ratio or 0.0
-        if myresults.get("unk") is None or unknown < myresults["unk"]:
-            myresults["unk"] = unknown
-    results = sorted(myresults.items(), key=itemgetter(1), reverse=True)
+    """Determine which proportion of the text is in the target language(s)."""
+    results = LanguageDetector(
+        dictionary_factory, token_sampler
+    ).get_text_percentage_in_each_language(text, lang, greedy)
     # post-processing
-    if len(results) > 1:
-        # switch unknown to the end
-        for i, item in enumerate(results):
-            if item[0] == "unk":
-                pair = results.pop(i)
-                results.append(pair)
-        # in case of ex-aequo use other token sampling to discriminate
-        if not greedy and results[0][1] == results[1][1]:
-            results = lang_detector(
-                text, lang=lang, greedy=True, token_sampler=backup_sampler
-            )
-    return results
+    if len(results) == 1:
+        return _convert_results_to_sorted_list(results)
+    # in case of ex-aequo
+    list_results = _convert_results_to_sorted_list(results)
```

Comment:
In some cases that's three conversions in total, maybe there's a better way.

Reply:
I've reduced the conversions to a minimum